src/grammar/expressions.rs

```diff
@@ -121,32 +134,64 @@ fn add_expr_node(p: &mut Parser, checkpoint: Checkpoint, sub_expr: Option<u8>) {
 fn prefix_bp(op: TokenKind) -> Option<(u8, Option<SyntaxKind>)> {
     match op {
         TokenKind::NotKw => Some((5, Some(SyntaxKind::LogicOp))),
-        TokenKind::Plus | TokenKind::Minus => Some((15, None)),
+        TokenKind::Plus | TokenKind::Minus => Some((17, None)),
         _ => None,
     }
 }
 
 fn postfix_bp(op: TokenKind) -> Option<(u8, Option<SyntaxKind>)> {
     match op {
-        TokenKind::Exclam => Some((17, None)),
+        TokenKind::Exclam => Some((19, None)),
         _ => None,
     }
 }
 
-fn infix_bp(op: TokenKind) -> Option<((u8, u8), Option<SyntaxKind>)> {
+#[allow(clippy::type_complexity)]
+fn infix_bp(
+    op: TokenKind,
+) -> Option<(
+    (u8, u8),
+    Option<SyntaxKind>,
+    Option<&'static dyn Fn(&mut Parser, u8)>,
+)> {
     match op {
-        TokenKind::OrKw => Some(((1, 2), Some(SyntaxKind::LogicOp))),
-        TokenKind::AndKw => Some(((3, 4), Some(SyntaxKind::LogicOp))),
-        TokenKind::ComparisonOp => Some(((7, 8), None)),
-        TokenKind::LikeKw => Some(((9, 10), None)),
-        TokenKind::Plus | TokenKind::Minus => Some(((11, 12), None)),
+        TokenKind::OrKw => Some(((1, 2), Some(SyntaxKind::LogicOp), None)),
```

Comment:
Do these values represent operand precedence? It's quite difficult to understand what the numbers specifically mean on their own, as these are not limited to this function only. I would suggest to comment this. If this precedence logic becomes more significant or is used in other places as well, introducing its own type might be an option, for example:

```rust
#[repr(u8)]
enum OpPrecedence {
    Or,
    And,
    ...
}
```

This kind of enum could be declared in a way that the order of variants can determine the weight / precedence of the operation. It also allows to implement [`PartialOrd`](https://doc.rust-lang.org/std/cmp/trait.PartialOrd.html) for comparison later. Not sure if this is relevant, as the parser only constructs the AST but does not execute it.

Reply:
I am not sure if enums can represent the precedence, given that there are operators that have the same binding power. Furthermore, some operators may have mappings for their AST node (see `TokenKind::OrKw` being mapped to `SyntaxKind::LogicOp`), and/or a callback (see `LIKE`, `BETWEEN` and `IN`).

On another note, would there be any "automatic" way to infer which `TokenKind` enum variant maps to which `OpPrecedence` variant?

Reply:
Hm, the way Rust Analyzer handles this, it's using the [`current_op`](https://github.com/rust-lang/rust-analyzer/blob/master/crates/parser/src/grammar/expressions.rs#L195) function which binds precedence to a single token or sequence of tokens, e.g. `<`, `<=`.

Reply:
> On another note, would there be any "automatic" way to infer which `TokenKind` enum variant maps to which `OpPrecedence` variant?

Not in an automated way (for example using variant values). They are handled as different types, but usually it's possible to provide a conversion function or trait implementation, e.g. [`TryFrom`](https://doc.rust-lang.org/stable/std/convert/trait.TryFrom.html).

Assume there exists a `TokenKind` & we would like to map that to op precedence, an example may look like:

```rust
#[derive(logos::Logos)]
enum TokenKind {
    #[regex("[ \t\n\r]+")]
    Whitespace,
    #[token("or", ignore(case))]
    OrKw,
    // ...
}

impl TokenKind {
    pub fn precedence(&self) -> Option<u16> {
        match self {
            Self::OrKw => 1,
            _ => None,
        }
    }
}
```

The method could also return a single `u16` with one specific value being the default (`0` or `std::u16::MAX`) but it depends on usage.

I'm not sure an extra `enum` for precedence adds much here, basically it is only there to determine the order in which operations are applied.

Most likely a single point in code needs to check the precedence of two tokens, then this might be helpful.

```rust
match (t1.precedence(), t2.precedence()) {
    (Some(p1), Some(p2)) => { ... } // compare both
    (_, _) => {} // ignore?
}
```